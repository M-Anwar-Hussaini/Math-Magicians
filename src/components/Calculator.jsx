import React, { useState } from 'react';
import Button from './Button';
import CalculationBar from './CalculationBar';
import calculate from '../logic/calculate';

const style = {
  width: '300px',
  margin: '32px auto',
};

const symbols = ['AC', '+/-', '%', 'รท', '7', '8', '9', 'x', '4', '5', '6', '-', '1', '2', '3', '+', '0', '.', '='];

const rightButtons = ['รท', 'x', '-', '+', '='];
const initObject = {
  total: null,
  next: null,
  operation: null,
};

const Calculator = () => {
  const [currentData, refactorCurrentDate] = useState(initObject);
  const { total, next } = currentData;
  const displayValue = next || total || 0;

  const btnClickHandler = (value) => {
    refactorCurrentDate((curData) => calculate(curData, value));
  };

  // Button Elements
  const elements = symbols.map((btnName) => {
    let btnType = 'outline-success';
    if (btnName === 'AC') {
      btnType = 'danger';
    } else if (rightButtons.includes(btnName)) {
      btnType = 'success';
    }
    return <Button value={btnName} col={btnName === '0' ? 6 : 3} key={btnName} btnType={btnType} click={() => btnClickHandler(btnName)} />;
  });

  return (
    <div className="row border border-success p-2 rounded" style={style}>
      <CalculationBar value={displayValue} />
      {elements}
    </div>
  );
};

export default Calculator;
