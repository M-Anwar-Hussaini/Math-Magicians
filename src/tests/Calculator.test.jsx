import '@testing-library/jest-dom/extend-expect';
import {fireEvent, render} from '@testing-library/react';
import Calculator from '../components/Calculator';

describe('Calculate', () => {
  it('renders Calculate component', () => {
    const calculator = render(<Calculator />);
    
    expect(calculator.container).toMatchSnapshot();
    
  })

  it('displaying button by clicking event', () => {
    const calc = render (<Calculator />);
    const display = calc.container.querySelector('.form-control');

    fireEvent.click(calc.getByText('9'));

    expect(display.innerHTML).toEqual('')
  })


})