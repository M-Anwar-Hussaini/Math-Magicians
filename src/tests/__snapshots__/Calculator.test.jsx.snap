// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculate renders Calculate component 1`] = `
<div>
  <div
    class="border border-success p-2 rounded box"
  >
    <input
      class="col-12 form-control mb-2 text-end"
      disabled=""
      readonly=""
      type="text"
      value="0"
    />
    <button
      class="border col-3 btn btn-danger"
      type="button"
    >
      AC
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      +/-
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      %
    </button>
    <button
      class="border col-3 btn btn-success"
      type="button"
    >
      รท
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      7
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      8
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      9
    </button>
    <button
      class="border col-3 btn btn-success"
      type="button"
    >
      x
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      4
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      5
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      6
    </button>
    <button
      class="border col-3 btn btn-success"
      type="button"
    >
      -
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      1
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      2
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      3
    </button>
    <button
      class="border col-3 btn btn-success"
      type="button"
    >
      +
    </button>
    <button
      class="border col-6 btn btn-outline-success"
      type="button"
    >
      0
    </button>
    <button
      class="border col-3 btn btn-outline-success"
      type="button"
    >
      .
    </button>
    <button
      class="border col-3 btn btn-success"
      type="button"
    >
      =
    </button>
  </div>
</div>
`;
